test/unit/qualityTest.test.js

```diff
@@ -0,0 +1,214 @@
+    //console.log(`Trial # ${i + 1}`);
+    trials.push(await trial(trainSize, testSize, createTrainerFn(), labelFn));
+  }
+
+  const averagedConfusionMatrix = {};
+  const keys = ['truePos', 'falsePos', 'falseNeg', 'trueNeg', 'precision', 'recall', 'accuracy'];
+
+  for (const key of keys) {
+    //console.log(`${key} ${JSON.stringify(resultsForCurrentNumberOfTrials.map(result => result[key]))}`)
+    const values = trials.map(t => t[key]);
+    const avg = average(values);
+    averagedConfusionMatrix[key] = avg;
+  }
+
+  return averagedConfusionMatrix;
+};
+
+const createConfusionMatrix = function(predictedOcean, numTrained, labelFn) {
+  //const confusionMatrix = {truePos: 0, falsePos: 0, falseNeg: 0, trueNeg: 0}
+  var truePos, falsePos, falseNeg, trueNeg;
+  truePos = falsePos = falseNeg = trueNeg = 0;
+
+  for (const fish of predictedOcean) {
+    const actualLabel = labelFn(fish);
+    const prediction = fish.result.predictedClassId;
+
+    if (prediction && actualLabel) {
+      truePos++;
+    } else if (prediction && !actualLabel) {
+      falsePos++;
+    } else if (!prediction && actualLabel) {
+      falseNeg++;
+    } else {
+      trueNeg++;
+    }
+  };
+
+  const totalSize = truePos + falsePos + falseNeg + trueNeg;
+
+  const precision = truePos / (truePos + falsePos);
+  const recall = truePos / (truePos + falseNeg);
+  const accuracy = (truePos + trueNeg) / totalSize;
+
+  const confusionMatrix = {
+    truePos: truePos,
+    falsePos: falsePos,
+    falseNeg: falseNeg,
+    trueNeg: trueNeg,
+    // If the model didn't predict "yes" for anything in the test data, precision is NaN.
+    // We can assume there was at least one positive in the test data and call this 0 precision.
+    precision: precision || 0,
+    recall: recall,
+    accuracy: accuracy
+  };
+
+  return confusionMatrix;
+};
+
+const analyzeConfusionMatrix = function(numTrained, cMatrix) {
+  console.log(`Num Trained: ${numTrained} Precision: ${cMatrix.precision.toFixed(2)} Recall: ${cMatrix.recall.toFixed(2)} Accuracy: ${cMatrix.accuracy.toFixed(2)}\n` +
+    `True Pos: ${cMatrix.truePos.toFixed(2)}\tFalse Pos: ${cMatrix.falsePos.toFixed(2)}\n` +
+    `False Neg: ${cMatrix.falseNeg.toFixed(2)} True Neg: ${cMatrix.trueNeg.toFixed(2)}\n`);
+};
+
+const floatEquals = (a, b) => {
+  return Math.abs(a - b) <= .0001;
+};
+
+const PartKey = Object.freeze({
+  BODY: 'body',
+  EYE: 'eye',
+  MOUTH: 'mouth',
+  PECTORAL_FIN_FRONT: 'pectoralFinFront',
+  DORSAL_FIN: 'dorsalFin',
+  TAIL: 'tail',
+  COLOR: 'colorPalette'
+});
+
+const NUM_TRIALS = 5;
+const TRAIN_SIZE = 100;
+
+describe('Model quality test', () => {
+  beforeAll(() => {
+    initFishData();
+  });
+
+  test('Round fish quality test', async () => {
+    const roundFishFn = (fish) => fish.body.knnData[1] === 0 ? 1 : 0;
+    const trainSize = TRAIN_SIZE;
+
+    const result = await performTrials({numTrials: NUM_TRIALS, trainSize: trainSize, testSize: 100, labelFn: roundFishFn});
```

Comment:
nit: should be broken up into multiple lines (there are a few of these in this file)

Reply:
👍